tests/code_update.py

```diff
@@ -0,0 +1,238 @@
+            "sign",
+            "-e",
+            f"{args.patched_file_name}.so",
+            "-c",
+            args.oeconfpath,
+            "-k",
+            args.oesignkeypath,
+        ],
+        capture_output=True,
+        check=True,
+    )
+
+    return get_code_id(f"{args.patched_file_name}.so.signed")
+
+
+def vote_to_accept(primary, proposal_id):
+    # vote to accept the new code id
+    result = infra.proc.ccall(
+        "./memberclient",
+        "vote",
+        "--accept",
+        "--cert=member1_cert.pem",
+        "--privk=member1_privk.pem",
+        f"--host={primary.host}",
+        f"--port={primary.tls_port}",
+        f"--id={proposal_id}",
+        "--ca=networkcert.pem",
+        "--sign",
+    )
+    j_result = json.loads(result.stdout)
+    assert not j_result["result"]
+
+    result = infra.proc.ccall(
+        "./memberclient",
+        "vote",
+        "--accept",
+        "--cert=member2_cert.pem",
+        "--privk=member2_privk.pem",
+        f"--host={primary.host}",
+        f"--port={primary.tls_port}",
+        f"--id={proposal_id}",
+        "--ca=networkcert.pem",
+        "--sign",
+    )
+    j_result = json.loads(result.stdout)
+    assert j_result["result"]
+
+
+def add_new_code(primary):
+    new_code_id = create_new_code_version(primary)
+    LOG.debug(f"New code id: {new_code_id}")
+
+    # first propose adding the new code id
+    result = infra.proc.ccall(
+        "./memberclient",
+        "add_code",
+        "--cert=member1_cert.pem",
+        "--privk=member1_privk.pem",
+        f"--host={primary.host}",
+        f"--port={primary.tls_port}",
+        "--ca=networkcert.pem",
+        f"--new_code_id={new_code_id}",
+    )
+
+    vote_to_accept(primary, 0)
+
+
+def create_node_using_new_code(network, args, node_id):
+    # add a node using unsupported code
+    assert network.create_and_add_node(
+        args.patched_file_name, args, node_id, False
+    ) == (False, infra.jsonrpc.ErrorCode.CODE_ID_NOT_FOUND)
+
+
+def run(args):
+    hosts = ["localhost", "localhost"]
+
+    with infra.ccf.network(
+        hosts, args.build_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
+    ) as network:
+        primary, others = network.start_and_join(args)
+
+        forwarded_args = {
+            arg: getattr(args, arg) for arg in infra.ccf.Network.node_args_to_forward
+        }
+
+        res, new_node, new_node_id = network.create_and_add_node(
+            args.package, args, 2, True
+        )
+        new_node.join_network()
+
+        # try to add a node using unsupported code
+        assert network.create_and_add_node(args.patched_file_name, args, 3, False) == (
+            False,
+            infra.jsonrpc.ErrorCode.CODE_ID_NOT_FOUND,
+        )
+
+        add_new_code(primary)
+
+        with open("networkcert.pem", mode="rb") as file:
+            net_cert = list(file.read())
+
+        new_nodes = set()
+        # add nodes using the same code id that failed earlier
```

Comment:
Is this comment still valid?

Reply:
I believe it is, do you find it to be redundant?

Reply:
The previous code used for the failed node was `3` but it seems that we start the range here from `4`.